File: 539.minimum-time-difference.java
public class Solution {
    public int findMinDifference(List<String> timePoints) {
        int n = timePoints.size();
        List<Time> times = new ArrayList<>();
        for (String tp : timePoints) {
            String[] strs = tp.split(":");
            times.add(new Time(Integer.parseInt(strs[0]), Integer.parseInt(strs[1])));
        }
        Collections.sort(times);
        Time earlist = times.get(0);
        times.add(new Time(earlist.h + 24, earlist.m));
        int minDiff = Integer.MAX_VALUE;
        for (int i = 0; i < n; i++) {
            int diff = (int) Math.abs(times.get(i).getDiff(times.get(i + 1)));
            minDiff = Math.min(minDiff, diff);
        }
        return minDiff;
    }
    
}

class Time implements Comparable<Time> {
    int h;
    int m;
    public Time(int h, int m) {
        this.h = h;
        this.m = m;
    }
    
    public int compareTo(Time other) {
        if (this.h == other.h) {
            return this.m - other.m;
        }
        return this.h - other.h;
    }
    
    public int getDiff(Time other) {
        return (this.h - other.h) * 60 + (this.m - other.m);            
    }
}

File: 539.minimum-time-difference.js
/*
 * @lc app=leetcode id=539 lang=javascript
 *
 * [539] Minimum Time Difference
 */

// @lc code=start
/**
 * @param {string[]} timePoints
 * @return {number}
 */
var findMinDifference = function (timePoints) {
  let mark = new Array(24 * 60).fill(0);
  for (let time of timePoints) {
    let t = time.split(":");
    let h = parseInt(t[0]);
    let m = parseInt(t[1]);
    if (mark[h * 60 + m]) {
      return 0;
    }
    mark[h * 60 + m] = true;
  }

  let prev = 0,
    min = Number.MAX_VALUE;
  let first = Number.MAX_VALUE,
    last = Number.MIN_VALUE;
  for (let i = 0; i < 24 * 60; i++) {
    if (mark[i]) {
      if (first != Number.MAX_VALUE) {
        min = Math.min(min, i - prev);
      }
      first = Math.min(first, i);
      last = Math.max(last, i);
      prev = i;
    }
  }

  min = Math.min(min, 24 * 60 - last + first);

  return min;
};
// @lc code=end
